### PR TITLE
[Ready]Moves traits to datums

### PR DESCRIPTION
**Pull request by XDTM**
I've been in a few situations where traits would be handy to track certain object or datum flags, and there's no particular reason that ties them to living mobs aside from being initially a disability rework.

### CODE REVIEW

**Comment by Cyberboss** [comment]
You seem to be stepping on component territory here. That's the general way we add properties/behaviors to datums

**Comment by XDTM** [comment]
My reasoning is that components are useful for complex behaviours, but unnecessarily heavy for flag-like properties.

**Comment by XDTM** [comment]
By the way, aren't mobs/objs supposed to inherit from datums? Do i need to copy the procs to each base type?

**Comment by nichlas0010** [comment]
>By the way, aren't mobs/objs supposed to inherit from datums? Do i need to copy the procs to each base type?

Mobs/objs inherit from /atom/, which stands for Area, Turf, Object, and Mob, AKA the four types that inherit from it. The opposite of the atoms are the /datum/s, which don't inherit from /atom/, and from who the /atom/s don't inherit from.

**Comment by Cyberboss** [comment]
In regards to flag-like properties datums/atoms/movables already have their own flag sets which we've been aiming to keep clear as possible, each probably has 20 or so more flags available. If something doesn't fit in the context of a flag, it gets promoted to a component.

**Comment by XDTM** [comment]
Traits are supposed to be used as flag substitutes in situations where multiple sources cause issues.

If, for example, i wanted to make an UNDETECTABLE flag for diseases, and two different symptoms add/remove it when present, you'd have an issue if you add both then remove one, as the flag would be deactivated despite the first symptom still being there.

Is it really better to make a whole component for the undetectable property here? Or copypasting the trait logic for each datum flag with these circumstances?

**Comment by Razharas** [comment]
You do it by having a special stage at addition/removal in which you make new trait check all other traits and see if they are ok with eachother and what to do if not and there is a special case
Just making it bit flags wouldnt solve it

**Comment by XDTM** [comment]
And i would have to do that for every single true/false property that has more than one source, which is why i'd prefer having a central helper proc instead, which also does the job better. I mean, look at all the traits used at living level; if used for objects or datums i bet they'd end up being very handy.

**Comment by fludd12** [comment]
As an example of XDTMs point, someone way in the past switched from a bitflag to a trait for blindness so you didn't regain sight by wearing and taking off a blindfold. You can't use flags for these, unless you want to make an unnecessarily large amount of flags. Making them components *might* work well, but I don't know the mem cost of having one for everything we use a trait for.

**Comment by SpaceManiac** [comment]
stale: doesn't compile and hasn't been touched in over two weeks

**Comment by XDTM** [comment]
i'll get it to compile and reopen it

**Comment by Razharas** [comment]
Famous last words

**Comment by ma44** [comment]
ma44 casts wand of healing at #38505

**Comment by XDTM** [comment]
@SpaceManiac PR resurrected, can be un-staled

File: code/datums/traits.dm [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		status_traits = list()
+		
+	if(!status_traits[trait])
+		status_traits[trait] = list(source)
+	else
+		status_traits[trait] |= list(source)
+		
+
+/datum/proc/remove_trait(trait, list/sources, force)
+	if(isnull(status_traits))
```
Comment (SpaceManiac) [inline code/datums/traits.dm]: Why create the list if it's empty, won't that just create more work for later calls of the same proc? Ditto the other three methods

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/datums/traits.dm (hunk 1 above)
Comment (XDTM) [reply code/datums/traits.dm]: oh yeah, you're right

**Pull request merged**